File: 725.split-linked-list-in-parts.py
#
# @lc app=leetcode id=725 lang=python3
#
# [725] Split Linked List in Parts
#


# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class LinkedList:
    def __init__(self):
        self.head = None
        self.size = 0
        self.tail = None

    def add(self, node):
        if self.head is None:
            self.head = node
            self.tail = node
            self.size = 1
        else:
            self.tail.next = node
            self.tail = self.tail.next
            self.size += 1


class Solution:
    def splitListToParts(
        self, head: Optional[ListNode], k: int
    ) -> List[Optional[ListNode]]:
        length = 0
        curr = head
        while curr:
            length += 1
            curr = curr.next

        sizes = [length // k] * k
        left = length % k
        idx = 0
        while left > 0:
            sizes[idx] += 1
            left -= 1
            idx += 1

        ans = []
        curr = head
        currGroupIdx = 0
        currLL = LinkedList()
        while curr:
            nextNode = curr.next
            curr.next = None
            if currLL.size == sizes[currGroupIdx]:
                ans.append(currLL.head)
                currLL = LinkedList()
                currGroupIdx += 1
            currLL.add(curr)
            curr = nextNode
        ans.append(currLL.head)
        while len(ans) < k:
            ans.append(None)
        return ans


# @lc code=end

File: 725.split-linked-list-in-parts.go
/*
 * @lc app=leetcode id=725 lang=golang
 *
 * [725] Split Linked List in Parts
 */

// @lc code=start
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func splitListToParts(head *ListNode, k int) []*ListNode {
	if head == nil {
		return make([]*ListNode, k)
	}
	// count and calculate for each sub linked lists
	n := 0
	curr := head
	for curr != nil {
		n++
		curr = curr.Next
	}
	nodes := make([]int, k)
	base := n / k
	remainder := n % k
	for i := 0; i < k; i++ {
		nodes[i] = base
		if remainder > 0 {
			nodes[i]++
			remainder--
		}
	}
	// prepare answer
	ans := make([]*ListNode, k)
	var prev *ListNode = nil
	curr = head
	for i := 0; i < k; i++ {
		ans[i] = curr
		x := nodes[i]
		// move curr x times
		for j := 0; j < x; j++ {
			if curr != nil {
				prev = curr
				curr = curr.Next
			}
		}
		// split
		prev.Next = nil
	}
	return ans
}

// @lc code=end
